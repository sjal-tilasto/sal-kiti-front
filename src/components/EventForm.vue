<template>
  <div>
    <b-row>
      <b-col>
        <h2 class="bg-sal-orange" v-if="edit">{{ $t("event.edit") }}</h2>
        <h2 class="bg-sal-orange" v-else>{{ $t("event.add") }}</h2>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-alert v-if="errors.main" variant="danger" show>
          <ul>
            <li v-for="e in errors.main" v-bind:key="e">{{ e }}</li>
          </ul>
        </b-alert>
      </b-col>
    </b-row>
    <b-row v-if="loadingEvent">
      <b-col>
        <b-spinner label="Loading..."></b-spinner>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <b-form @submit="onSubmit">
          <b-form-group
            id="input-group-organization"
            :label="$t('organizer')"
            label-for="input-organization"
          >
            <b-form-select
              id="input-organization"
              v-model="event.organization"
              :options="organizations"
              textField="name"
              valueField="id"
              required
            >
            </b-form-select>
            <b-form-invalid-feedback :state="(!'organization') in errors">
              <ul>
                <li v-for="e in errors.organization" v-bind:key="e">{{ e }}</li>
              </ul>
            </b-form-invalid-feedback>
          </b-form-group>
          <b-form-group
            id="input-group-name"
            :label="$t('name')"
            label-for="input-name"
          >
            <b-form-input
              id="input-name"
              v-model="event.name"
              required
            ></b-form-input>
            <b-form-invalid-feedback :state="(!'name') in errors">
              <ul>
                <li v-for="e in errors.name" v-bind:key="e">{{ e }}</li>
              </ul>
            </b-form-invalid-feedback>
          </b-form-group>
          <b-form-group
            id="input-group-description"
            :label="$t('description')"
            label-for="input-description"
            :description="$t('event.description_help')"
          >
            <b-form-textarea
              id="input-description"
              v-model="event.description"
              rows="3"
              max-rows="6"
            ></b-form-textarea>
            <b-form-invalid-feedback :state="(!'description') in errors">
              <ul>
                <li v-for="e in errors.description" v-bind:key="e">{{ e }}</li>
              </ul>
            </b-form-invalid-feedback>
          </b-form-group>
          <b-form-group
            id="input-group-location"
            :label="$t('location')"
            label-for="input-location"
          >
            <b-form-input
              id="input-location"
              v-model="event.location"
              required
            ></b-form-input>
            <b-form-invalid-feedback :state="(!'location') in errors">
              <ul>
                <li v-for="e in errors.location" v-bind:key="e">{{ e }}</li>
              </ul>
            </b-form-invalid-feedback>
          </b-form-group>
          <b-form-group
            id="input-group-date-start"
            :label="$t('date_start')"
            label-for="input-date-start"
          >
            <b-form-input
              id="input-date-start"
              v-model="event.date_start"
              type="date"
              required
            ></b-form-input>
            <b-form-invalid-feedback :state="(!'date_start') in errors">
              <ul>
                <li v-for="e in errors.date_start" v-bind:key="e">{{ e }}</li>
              </ul>
            </b-form-invalid-feedback>
          </b-form-group>
          <b-form-group
            id="input-group-date-end"
            :label="$t('date_end')"
            label-for="input-date-end"
          >
            <b-form-input
              id="input-date-end"
              v-model="event.date_end"
              type="date"
              required
            ></b-form-input>
            <b-form-invalid-feedback :state="(!'date_end') in errors">
              <ul>
                <li v-for="e in errors.date_end" v-bind:key="e">{{ e }}</li>
              </ul>
            </b-form-invalid-feedback>
          </b-form-group>
          <b-form-group
            id="input-group-optional_dates"
            :label="$t('event.optional_dates')"
            label-for="input-organization"
            :description="$t('event.optional_dates_help')"
          >
            <b-form-textarea
              id="input-optional_dates"
              v-model="event.optional_dates"
              rows="3"
              max-rows="6"
            >
            </b-form-textarea>
          </b-form-group>
          <b-form-group
            id="input-group-categories"
            :label="$t('event.categories')"
            label-for="input-categories"
            :description="$t('event.categories_help')"
          >
            <b-form-textarea
              id="input-categories"
              v-model="event.categories"
              rows="3"
              max-rows="6"
            >
            </b-form-textarea>
          </b-form-group>
          <b-form-group
            id="input-group-web_page"
            :label="$t('event.web_page')"
            label-for="input-web_page"
            :description="$t('event.web_page_help')"
          >
            <b-form-input id="input-web_page" v-model="event.web_page">
            </b-form-input>
            <b-form-invalid-feedback :state="(!'web_page') in errors">
              <ul>
                <li v-for="e in errors.web_page" v-bind:key="e">{{ e }}</li>
              </ul>
            </b-form-invalid-feedback>
          </b-form-group>
          <b-form-group
            id="input-group-invitation"
            :label="$t('event.invitation')"
            label-for="input-invitation"
            :description="$t('event.invitation_help')"
          >
            <b-form-input id="input-invitation" v-model="event.invitation">
            </b-form-input>
            <b-form-invalid-feedback :state="(!'invitation') in errors">
              <ul>
                <li v-for="e in errors.invitation" v-bind:key="e">{{ e }}</li>
              </ul>
            </b-form-invalid-feedback>
          </b-form-group>
          <b-form-group
            id="input-group-safety_plan"
            :label="$t('event.safety_plan')"
            label-for="input-safety_plan"
            :description="$t('event.safety_plan_help')"
          >
            <b-form-checkbox id="input-safety_plan" v-model="event.safety_plan">
            </b-form-checkbox>
          </b-form-group>
          <b-form-group
            id="input-group-notes"
            :label="$t('event.notes')"
            label-for="input-notes"
            :description="$t('event.notes_help')"
          >
            <b-form-textarea
              id="input-notes"
              v-model="event.notes"
              rows="3"
              max-rows="6"
            >
            </b-form-textarea>
          </b-form-group>
          <b-form-group
            id="input-group-toc_agreement"
            :label="$t('event.toc_agreement')"
            label-for="input-toc_agreement"
            :description="$t('event.toc_agreement_help')"
          >
            <b-form-checkbox
              id="input-toc_agreement"
              v-model="event.toc_agreement"
              name="toc-checkbox"
              required
            >
            </b-form-checkbox>
          </b-form-group>
          <p>
            <b-link :href="$t('event.toc_agreement_url')" target="_blank"
              >({{ $t("link") }})</b-link
            >
          </p>

          <div>
            <b-button
              type="submit"
              variant="light"
              class="btn-orange space-right"
            >
              <section v-if="edit">{{ $t("update") }}</section>
              <section v-else>{{ $t("create") }}</section>
            </b-button>
          </div>
          <br />
        </b-form>
      </b-col>
    </b-row>
    <b-row>
      <b-col>
        <EventFormContacts v-if="eventId" :eventId="eventId" />
      </b-col>
    </b-row>
  </div>
</template>

<script>
/**
 * Event create and edit form
 */
import { HTTP } from "../api/BaseApi.js";
import getCookie from "../utils/GetCookie";
import errorParser from "../utils/ErrorParser";
import apiGet from "../mixins/ApiGet";
import EventFormContacts from "@/components/EventFormContacts.vue";

export default {
  name: "EventForm",
  mixins: [apiGet],
  components: {
    EventFormContacts
  },
  data() {
    return {
      config: {
        headers: {
          "X-CSRFToken": getCookie("csrftoken")
        }
      },
      edit: false,
      errors: {},
      eventId: null,
      event: {
        date_end: null,
        date_start: null,
        location: null,
        name: null,
        description: "",
        organization: null,
        optional_dates: "",
        categories: "",
        notes: "",
        web_page: "",
        invitation: "",
        safety_plan: false,
        toc_agreement: false
      },
      loadingEvent: false,
      loadingOrganizations: false,
      organizations: []
    };
  },
  mounted() {
    this.getOrganizations(true, false, true);
    if (this.$route.params.event_id) {
      this.edit = true;
      this.getEvent(this.$route.params.event_id);
      this.eventId = this.$route.params.event_id.toString();
    }
  },
  methods: {
    /**
     * Create or edit event
     *
     * @param evt
     */
    onSubmit(evt) {
      evt.preventDefault();
      if (this.$route.params.event_id) {
        this.putEvent(this.$route.params.event_id);
      } else {
        this.postEvent();
      }
    },
    /**
     * Create a new event
     *
     * @returns {Promise<void>}
     */
    async postEvent() {
      this.erros = {};
      this.event.toc_agreement = true;
      HTTP.post("events/", this.event, this.config)
        .then((response) => {
          this.$router.push({
            name: "event",
            params: { event_id: response.data.id }
          });
        })
        .catch((error) => {
          this.errors = errorParser.form.bind(this)(error);
        });
    },
    /**
     * Edit event
     *
     * @param {number} id
     * @returns {Promise<void>}
     */
    async putEvent(id) {
      this.erros = {};
      this.event.toc_agreement = true;
      HTTP.put("events/" + id + "/", this.event, this.config)
        .then((response) => {
          this.$router.push({
            name: "event",
            params: { event_id: response.data.id }
          });
        })
        .catch((error) => {
          this.errors = errorParser.form.bind(this)(error);
        });
    }
  }
};
</script>

<style scoped></style>
