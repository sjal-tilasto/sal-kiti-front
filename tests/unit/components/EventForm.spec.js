import { createLocalVue, shallowMount } from "@vue/test-utils";
import EventForm from "@/components/EventForm.vue";
import BootstrapVue from "bootstrap-vue";
import Vuex from "vuex";
import flushPromises from "flush-promises";

const localVue = createLocalVue();
localVue.use(BootstrapVue);
localVue.use(Vuex);

const $router = {
  push: jest.fn()
};

const formData = {
  name: "Demo Champion",
  date_start: "2019-01-22",
  date_end: "2019-01-24",
  location: "Village Field",
  organization: 1
};

describe("EventForm.vue", () => {
  it("renders event form with even information", async () => {
    const $route = {
      params: { query: { event_id: 1 } }
    };
    const wrapper = shallowMount(EventForm, {
      localVue,
      mocks: {
        $route,
        $router
      }
    });
    await flushPromises();
    expect(wrapper.element).toMatchSnapshot();
  });

  it("posts new event", async () => {
    const $route = {
      params: { query: {} }
    };
    const wrapper = shallowMount(EventForm, {
      localVue,
      mocks: {
        $route,
        $router
      }
    });
    wrapper.setData({
      event: formData
    });
    await wrapper.vm.postEvent();
    expect($router.push).toHaveBeenCalledWith({
      name: "event",
      params: { event_id: 5 }
    });
  });

  it("updates event data", async () => {
    const $route = {
      params: { query: { event_id: 1 } }
    };
    const wrapper = shallowMount(EventForm, {
      localVue,
      mocks: {
        $route,
        $router
      }
    });
    wrapper.vm.$data.event.name = "Village Champs";
    await wrapper.vm.putEvent(1);
    expect($router.push).toHaveBeenCalledWith({
      name: "event",
      params: { event_id: 1 }
    });
  });
});
