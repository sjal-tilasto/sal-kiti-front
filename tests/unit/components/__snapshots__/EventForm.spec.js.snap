// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventForm.vue renders event form with even information 1`] = `
<div>
  <b-row-stub
    tag="div"
  >
    <b-col-stub
      tag="div"
    >
      <h2
        class="bg-sal-orange"
      >
        event.add
      </h2>
    </b-col-stub>
  </b-row-stub>
   
  <b-row-stub
    tag="div"
  >
    <b-col-stub
      tag="div"
    >
      <!---->
    </b-col-stub>
  </b-row-stub>
   
  <!---->
   
  <b-row-stub
    tag="div"
  >
    <b-col-stub
      tag="div"
    >
      <b-form-stub>
        <b-form-group-stub
          feedbackarialive="assertive"
          id="input-group-organization"
          label="organizer"
          labelfor="input-organization"
        >
          <b-form-select-stub
            disabledfield="disabled"
            htmlfield="html"
            id="input-organization"
            labelfield="label"
            options="[object Object],[object Object],[object Object]"
            optionsfield="options"
            required="true"
            selectsize="0"
            textfield="name"
            valuefield="id"
          />
        </b-form-group-stub>
         
        <b-form-group-stub
          feedbackarialive="assertive"
          id="input-group-name"
          label="name"
          labelfor="input-name"
        >
          <b-form-input-stub
            debounce="0"
            id="input-name"
            required="true"
            type="text"
          />
        </b-form-group-stub>
         
        <b-form-group-stub
          description="event.description_help"
          feedbackarialive="assertive"
          id="input-group-description"
          label="description"
          labelfor="input-description"
        >
          <b-form-textarea-stub
            debounce="0"
            id="input-description"
            maxrows="6"
            rows="3"
            value=""
            wrap="soft"
          />
        </b-form-group-stub>
         
        <b-form-group-stub
          feedbackarialive="assertive"
          id="input-group-location"
          label="location"
          labelfor="input-location"
        >
          <b-form-input-stub
            debounce="0"
            id="input-location"
            required="true"
            type="text"
          />
        </b-form-group-stub>
         
        <b-form-group-stub
          feedbackarialive="assertive"
          id="input-group-date-start"
          label="date_start"
          labelfor="input-date-start"
        >
          <b-form-input-stub
            debounce="0"
            id="input-date-start"
            required="true"
            type="date"
          />
        </b-form-group-stub>
         
        <b-form-group-stub
          feedbackarialive="assertive"
          id="input-group-date-end"
          label="date_end"
          labelfor="input-date-end"
        >
          <b-form-input-stub
            debounce="0"
            id="input-date-end"
            required="true"
            type="date"
          />
        </b-form-group-stub>
         
        <b-form-group-stub
          description="event.optional_dates_help"
          feedbackarialive="assertive"
          id="input-group-optional_dates"
          label="event.optional_dates"
          labelfor="input-organization"
        >
          <b-form-textarea-stub
            debounce="0"
            id="input-optional_dates"
            maxrows="6"
            rows="3"
            value=""
            wrap="soft"
          />
        </b-form-group-stub>
         
        <b-form-group-stub
          description="event.categories_help"
          feedbackarialive="assertive"
          id="input-group-categories"
          label="event.categories"
          labelfor="input-categories"
        >
          <b-form-textarea-stub
            debounce="0"
            id="input-categories"
            maxrows="6"
            rows="3"
            value=""
            wrap="soft"
          />
        </b-form-group-stub>
         
        <b-form-group-stub
          description="event.web_page_help"
          feedbackarialive="assertive"
          id="input-group-web_page"
          label="event.web_page"
          labelfor="input-web_page"
        >
          <b-form-input-stub
            debounce="0"
            id="input-web_page"
            type="text"
            value=""
          />
        </b-form-group-stub>
         
        <b-form-group-stub
          description="event.invitation_help"
          feedbackarialive="assertive"
          id="input-group-invitation"
          label="event.invitation"
          labelfor="input-invitation"
        >
          <b-form-input-stub
            debounce="0"
            id="input-invitation"
            type="text"
            value=""
          />
        </b-form-group-stub>
         
        <b-form-group-stub
          description="event.safety_plan_help"
          feedbackarialive="assertive"
          id="input-group-safety_plan"
          label="event.safety_plan"
          labelfor="input-safety_plan"
        >
          <b-form-checkbox-stub
            id="input-safety_plan"
          />
        </b-form-group-stub>
         
        <b-form-group-stub
          description="event.notes_help"
          feedbackarialive="assertive"
          id="input-group-notes"
          label="event.notes"
          labelfor="input-notes"
        >
          <b-form-textarea-stub
            debounce="0"
            id="input-notes"
            maxrows="6"
            rows="3"
            value=""
            wrap="soft"
          />
        </b-form-group-stub>
         
        <b-form-group-stub
          description="event.toc_agreement_help"
          feedbackarialive="assertive"
          id="input-group-toc_agreement"
          label="event.toc_agreement"
          labelfor="input-toc_agreement"
        >
          <b-form-checkbox-stub
            id="input-toc_agreement"
            name="toc-checkbox"
            required="true"
          />
        </b-form-group-stub>
         
        <p>
          <b-link-stub
            event="click"
            href="event.toc_agreement_url"
            routertag="a"
            target="_blank"
          >
            (link)
          </b-link-stub>
        </p>
         
        <div>
          <b-button-stub
            class="btn-orange space-right"
            event="click"
            routertag="a"
            tag="button"
            target="_self"
            type="submit"
            variant="light"
          >
            <section>
              create
            </section>
          </b-button-stub>
        </div>
         
        <br />
      </b-form-stub>
    </b-col-stub>
  </b-row-stub>
   
  <b-row-stub
    tag="div"
  >
    <b-col-stub
      tag="div"
    >
      <!---->
    </b-col-stub>
  </b-row-stub>
</div>
`;
